### PR TITLE
[compiler-rt][rtsan] Introduce first end to end RTsan lit tests, enable instrumented unit tests

### PR DESCRIPTION
**Pull request by cjappl**
We can run our full end to end test suite now that #102622 has been merged

### CODE REVIEW

**Comment by cjappl** [comment]
(trying some new reviewers I saw were on the codeowners group for compiler-rt sanitizers, just let me know if you would not like to be requested in the future!)

**Comment by cjappl** [comment]
CC for review @davidtrevelyan

File: compiler-rt/test/rtsan/test_rtsan.cpp [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// RUN: %clangxx -fsanitize=realtime %s -o %t
```
Comment (MaskRay) [inline compiler-rt/test/rtsan/test_rtsan.cpp:1]: The directory is about test/rtsan, so any word like "test" "rtsan" does not carry more information.

You can call it `basic.cpp` if you think the readers can start reading this test first.

File: compiler-rt/test/rtsan/test_rtsan.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+// RUN: not %run %t 2>&1 | FileCheck %s
+// UNSUPPORTED: ios
+
+// Intent: Ensure that an intercepted call in a [[clang::nonblocking]] function
+//         is flagged as an error. Basic smoke test.
+
+#include <stdlib.h>
+
+void violation() [[clang::nonblocking]] { void *Ptr = malloc(2); }
```
Comment (MaskRay) [inline compiler-rt/test/rtsan/test_rtsan.cpp:10]: perhaps print it or store this to a global variable to ensure that optimizers don't optimize this out.

`ptr`

**Review by vitalybuka — APPROVED** [review]

**Review by vitalybuka — APPROVED** [review]

**Comment by vitalybuka** [comment]
Please fix these warnings, or some bots will fail:

rtsan_test_functional.cpp:159:44: warning: unused variable 'x'
rtsan_test_functional.cpp:163:28: warning: unused variable 'x'

gtest.h:1379:11: warning: comparison of integers of different signs
  from rtsan_test_functional.cpp:148:5:

rtsan_test_interceptors.cpp:325:59: warning: variable 'thread_info' is uninitialized when used here

**Comment by cjappl** [comment]
> Please fix these warnings, or some bots will fail:
> 
> rtsan_test_functional.cpp:159:44: warning: unused variable 'x' rtsan_test_functional.cpp:163:28: warning: unused variable 'x'
> 
> gtest.h:1379:11: warning: comparison of integers of different signs from rtsan_test_functional.cpp:148:5:
> 
> rtsan_test_interceptors.cpp:325:59: warning: variable 'thread_info' is uninitialized when used here

Thank you for catching these @vitalybuka! That saved me a lot of headache. 

Did you pull these warnings from somewhere specific, or just building on your own machine? I want to know so I can catch these before I put up any future reviews, and I don't see them on my local mac or ubuntu builds (or the build machine logs)

**Comment by cjappl** [comment]
Wait, scratch that, they did turn up!! I will ensure they are fixed in just a sec

**Comment by cjappl** [comment]
I am in the process of reverting this test for the time being to buy myself time to investigate further, will update this comment with the reverted commit

**Comment by cjappl** [comment]
Disabled in 11ba2eee59c6, #106079

**Comment by Prabhuk** [comment]
Hi! This patch seems to break our toolchain builders.

```
FAILED: compiler-rt/lib/rtsan/tests/Rtsan-x86_64-Test /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/lib/rtsan/tests/Rtsan-x86_64-Test 
cd /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/lib/rtsan/tests && /b/s/w/ir/x/w/llvm_build/./bin/clang++ RtsanTestObjects.rtsan_test_functional.cpp.x86_64.o RtsanTestObjects.rtsan_test_interceptors.cpp.x86_64.o RtsanTestObjects.rtsan_test_main.cpp.x86_64.o RtsanTestObjects.gtest-all.cc.x86_64.o RtsanTestObjects.gmock-all.cc.x86_64.o -o /b/s/w/ir/x/w/llvm_build/runtimes/runtimes-x86_64-unknown-linux-gnu-bins/compiler-rt/lib/rtsan/tests/./Rtsan-x86_64-Test -fuse-ld=lld -fuse-ld=lld -Wl,--color-diagnostics -Wthread-safety -Wthread-safety-reference -Wthread-safety-beta -resource-dir=/b/s/w/ir/x/w/llvm_build/./lib/../lib/clang/20 --unwindlib=none -nostdlib++ /b/s/w/ir/x/w/llvm_build/lib/x86_64-unknown-linux-gnu/libunwind.a /b/s/w/ir/x/w/llvm_build/lib/x86_64-unknown-linux-gnu/libc++.a -no-pie -ldl -lrt -lm -pthread -fsanitize=realtime
ld.lld: error: undefined symbol: __atomic_is_lock_free
>>> referenced by atomic_base.h:40 (/b/s/w/ir/x/w/llvm_build/include/c++/v1/__atomic/atomic_base.h:40)
>>>               RtsanTestObjects.rtsan_test_functional.cpp.x86_64.o:(TestRtsan_AccessingALargeAtomicVariableDiesWhenRealtime_Test::TestBody())

ld.lld: error: undefined symbol: __atomic_load
>>> referenced by cxx_atomic_impl.h:317 (/b/s/w/ir/x/w/llvm_build/include/c++/v1/__atomic/cxx_atomic_impl.h:317)
>>>               RtsanTestObjects.rtsan_test_functional.cpp.x86_64.o:(TestRtsan_AccessingALargeAtomicVariableDiesWhenRealtime_Test::TestBody())
>>> referenced by cxx_atomic_impl.h:317 (/b/s/w/ir/x/w/llvm_build/include/c++/v1/__atomic/cxx_atomic_impl.h:317)
>>>               RtsanTestObjects.rtsan_test_functional.cpp.x86_64.o:(TestRtsan_AccessingALargeAtomicVariableDiesWhenRealtime_Test::TestBody())
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
[19
```

Link to our full build log file: https://logs.chromium.org/logs/fuchsia/buildbucket/cr-buildbucket/8738529178618519553/+/u/clang/test/stdout

**Comment by Prabhuk** [comment]
Some observations from our team in our early investigation:

- libclang_rt.atomic.so is not linked in for the test binaries.
- We don't see `-latomic` in the link even though cmake file for the test seems to try to set it: https://github.com/llvm/llvm-project/blob/0359b9a2306566a872cede1fcdaa4edc34e9398e/compiler-rt/lib/rtsan/tests/CMakeLists.txt#L39

We suspect the problem here could be with detection of `COMPILER_RT_HAS_LIBATOMIC`.

**Comment by cjappl** [comment]
Hi @Prabhuk,

Thanks for the early legwork on this, and sorry for the breakage. I am more than happy to help you get things back to green.

> Some observations from our team in our early investigation:
> libclang_rt.atomic.so is not linked in for the test binaries.
>   We don't see -latomic in the link even though cmake file for the test seems to try to set it:

 >   [llvm-project/compiler-rt/lib/rtsan/tests/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/0359b9a2306566a872cede1fcdaa4edc34e9398e/compiler-rt/lib/rtsan/tests/CMakeLists.txt#L39)

>    Line 39 in [0359b9a](https://github.com/llvm/llvm-project/commit/0359b9a2306566a872cede1fcdaa4edc34e9398e)

>     append_list_if(COMPILER_RT_HAS_LIBATOMIC -latomic RTSAN_UNITTEST_LINK_FLAGS) 

> We suspect the problem here could be with detection of COMPILER_RT_HAS_LIBATOMIC.


This theory has a lot of weight for me, when I compile on ubuntu, for instance, and apply this patch:

```
diff --git a/compiler-rt/lib/rtsan/tests/CMakeLists.txt b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
index 0320bbad5921..c31370fa01be 100644
--- a/compiler-rt/lib/rtsan/tests/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
@@ -37,6 +37,7 @@ set(RTSAN_UNITTEST_LINK_FLAGS
   -no-pie)

 append_list_if(COMPILER_RT_HAS_LIBATOMIC -latomic RTSAN_UNITTEST_LINK_FLAGS)
+message(FATAL_ERROR "COMPILER_RT_HAS_LIBATOMIC: ${COMPILER_RT_HAS_LIBATOMIC}")
 append_list_if(COMPILER_RT_HAS_LIBDL -ldl RTSAN_UNITTEST_LINK_FLAGS)
 append_list_if(COMPILER_RT_HAS_LIBRT -lrt RTSAN_UNITTEST_LINK_FLAGS)
 append_list_if(COMPILER_RT_HAS_LIBM -lm RTSAN_UNITTEST_LINK_FLAGS)
```

I get, as expected:
```
-- For aarch64 builtins preferring aarch64/fp_mode.c to fp_mode.c
CMake Error at /test_radsan/llvm-project/compiler-rt/lib/rtsan/tests/CMakeLists.txt:40 (message):
  COMPILER_RT_HAS_LIBATOMIC: 1
```

Can you describe what OS and arch you are compiling for? (I see the build logs, but I don't want to assume anything). I typically run my tests locally against MacOS and ubuntu, which is why I may not have seen this error in my development.

If you explicitly hack to add the linker flag, such as:

```
diff --git a/compiler-rt/lib/rtsan/tests/CMakeLists.txt b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
index 0320bbad5921..b9755786965a 100644
--- a/compiler-rt/lib/rtsan/tests/CMakeLists.txt
+++ b/compiler-rt/lib/rtsan/tests/CMakeLists.txt
@@ -36,7 +36,7 @@ set(RTSAN_UNITTEST_LINK_FLAGS
   ${SANITIZER_TEST_CXX_LIBRARIES}
   -no-pie)

-append_list_if(COMPILER_RT_HAS_LIBATOMIC -latomic RTSAN_UNITTEST_LINK_FLAGS)
+append_list_if(ON -latomic RTSAN_UNITTEST_LINK_FLAGS)
 append_list_if(COMPILER_RT_HAS_LIBDL -ldl RTSAN_UNITTEST_LINK_FLAGS)
 append_list_if(COMPILER_RT_HAS_LIBRT -lrt RTSAN_UNITTEST_LINK_FLAGS)
 append_list_if(COMPILER_RT_HAS_LIBM -lm RTSAN_UNITTEST_LINK_FLAGS)
```

Does the flag show up, and it compiles? I think this would be another indication the detection of `COMPILER_RT_HAS_LIBATOMIC` is funky for this specific config.

**Comment by cjappl** [comment]
If you are in need of debugging very quickly, I am happy to talk on discord (@apple412) or email (in the commit messages), just shout if I can be of assistance.

**Comment by petrhosek** [comment]
We don't install `libatomic` on our builders (which is a part of libgcc), instead we use `libclang_rt.atomic` that comes directly from compiler-rt. I think we need to update the following logic to properly handle both `libatomic` and `libclang_rt.atomic`:

https://github.com/llvm/llvm-project/blob/b2dd8405d990e525e8bc823e5130eb096bd0cdd0/compiler-rt/cmake/config-ix.cmake#L21-L34

**Comment by petrhosek** [comment]
> We don't install `libatomic` on our builders (which is a part of libgcc), instead we use `libclang_rt.atomic` that comes directly from compiler-rt. I think we need to update the following logic to properly handle both `libatomic` and `libclang_rt.atomic`:
> 
> https://github.com/llvm/llvm-project/blob/b2dd8405d990e525e8bc823e5130eb096bd0cdd0/compiler-rt/cmake/config-ix.cmake#L21-L34

https://github.com/llvm/llvm-project/pull/106603 implements support for using compiler-rt atomic library.

**Pull request merged**